contrib/src/msgpack/mod.rs

```diff
@@ -0,0 +1,132 @@
+/// ```rust,ignore
+/// #[post("/users/", format = "application/msgpack", data = "<user>")]
+/// fn new_user(user: MsgPack<User>) {
+///     ...
+/// }
+/// ```
+///
+/// You don't _need_ to use `format = "application/msgpack"`, but it _may_ be what
+/// you want. Using `format = application/msgpack` means that any request that
+/// doesn't specify "application/msgpack" as its first `Content-Type:` header
+/// parameter will not be routed to this handler. By default, Rocket will accept a
+/// Content Type of any of the following for MessagePack data:
+/// `application/msgpack`, `application/x-msgpack`, or `bin/msgpack`.
+///
+/// If you're responding with MessagePack data, return a `MsgPack<T>` type, where `T`
+/// implements `Serialize` from [Serde](https://github.com/serde-rs/serde). The
+/// content type of the response is set to `application/msgpack` automatically.
+///
+/// ```rust,ignore
+/// #[get("/users/<id>")]
+/// fn user(id: usize) -> MsgPack<User> {
+///     let user_from_id = User::from(id);
+///     ...
+///     MsgPack(user_from_id)
+/// }
+/// ```
+#[derive(Debug)]
+pub struct MsgPack<T>(pub T);
+
+impl<T> MsgPack<T> {
+    /// Consumes the `MsgPack` wrapper and returns the wrapped item.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use rocket_contrib::MsgPack;
+    /// let string = "Hello".to_string();
+    /// let my_msgpack = MsgPack(string);
+    /// assert_eq!(my_msgpack.into_inner(), "Hello".to_string());
+    /// ```
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+/// Maximum size of MessagePack data is 1MB.
+/// TODO: Determine this size from some configuration parameter.
+const MAX_SIZE: u64 = 1048576;
+
+impl<T: Deserialize> FromData for MsgPack<T> {
+    type Error = MsgPackError;
+
+    fn from_data(request: &Request, data: Data) -> data::Outcome<Self, Self::Error> {
+        // Accepted content types are:
+        // `application/msgpack`, `application/x-msgpack`, `bin/msgpack`, and `bin/x-msgpack`
+        if !request.content_type().map_or(false, |ct|
+            (ct.ttype == "application" || ct.ttype == "bin") &&
+            (ct.subtype == "msgpack" || ct.subtype == "x-msgpack")) {
+            error_!("Content-Type is not MessagePack.");
+            return Outcome::Forward(data);
+        }
+
+        let mut buf = Vec::new();
+        if let Err(e) = data.open().take(MAX_SIZE).read_to_end(&mut buf) {
+            let e = MsgPackError::InvalidDataRead(e);
+            error_!("Couldn't read request data: {:?}", e);
+            return Outcome::Failure((Status::BadRequest, e));
+        };
+
+        match rmp_serde::from_slice(&buf).map(|val| MsgPack(val)) {
+            Ok(value) => Outcome::Success(value),
+            Err(e) => {
+                error_!("Couldn't parse MessagePack body: {:?}", e);
+                Outcome::Failure((Status::BadRequest, e))
+            }
+        }
+    }
+}
+
+// Serializes the wrapped value into MessagePack. Returns a response with Content-Type
```

Comment:
This should be a real docstring, not a comment. Use three `///`. I've fixed this in the `JSON` docs.